enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,413 @@
+A) Setting HTTP headers by cluster admin via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variables, `ROUTER_SET_HTTP_HEADERS`, which specify the headers to set in the
+frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `ROUTER_SET_HTTP_HEADERS` are comma-delimited lists of base64 encoded values of
+the form `name:value` separated by comma.
+The template translates these values into an appropriate `http-response set-header` stanzas.
+For example, if `ROUTER_SET_HTTP_HEADERS` has encoded value `WC1GcmFtZS1PcHRpb25zOiBERU5Z,WC1YU1MtUHJvdGVjdGlvbjogMTttb2RlPWJsb2Nr` 
+of `X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block` then the template adds the following stanzas to
+the frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config:
+```conf
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting HTTP headers via Route's Spec by Users/Devops/Developers.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+
+The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .
+The enhancement adds a field in ServiceAliasConfig called `SetHTTPHeadersViaRoute`, which specify the headers to set in the
+backend section `Plain http backend or backend with TLS terminated at the edge or a secure backend with re-encryption.` of the
+haproxy.config at this [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609).
+The router sets this variable based on the Route's
+`spec.httpSetHeaders` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
+
+The values of `SetHTTPHeadersViaRoute` will be set in the same way as described in the section A) above for  `ROUTER_SET_HTTP_HEADERS`
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+- Do we need to set request headers ?
+- Do we need a delete option ?
```

Comment:
I think so:

> delete Server & X-Powered-By

https://issues.redhat.com/browse/RFE-464?focusedCommentId=20025699&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-20025699

> We will allow global replace, delete and append options.

https://issues.redhat.com/browse/RFE-464?focusedCommentId=20064927&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-20064927

Is there a good reason not to include a delete option?  

For what it's worth, Gateway API has a delete option: https://github.com/kubernetes-sigs/gateway-api/blob/e7260d356a01afc973772610b6223686a971b43e/apis/v1beta1/httproute_types.go#L778-L798

Reply:
changed